envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua

```diff
@@ -62,25 +62,30 @@ end
 
 function envoy_on_response(handle)
     local rbacMetadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.rbac") or {}
+    local is_shadow_denied = (rbacMetadata["shadow_engine_result"] or "") == "denied"
     local lua_metadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.lua") or {}
     local allowed_client = lua_metadata["request.info.allowed_client"] or false
 
-    if should_log_request(rbacMetadata, allowed_client) then
-        log_request(lua_metadata, handle)
+    if should_log_request(is_shadow_denied, allowed_client) then
+        local upstream_request_time = handle:headers():get("x-envoy-upstream-service-time")
+        local rbac_action = "shadow_denied"
+
+        if is_shadow_denied and upstream_request_time == nil then
+            rbac_action = "denied"
+        end
+        log_request(lua_metadata, handle, rbac_action)
     end
 end
 
-function should_log_request(rbacMetadata, allowed_client)
-    local is_shadow_denied = (rbacMetadata["shadow_engine_result"] or "") == "denied"
-
+function should_log_request(is_shadow_denied, allowed_client)
     if (not is_shadow_denied) and allowed_client == true then
         return true
+    else
+        return is_shadow_denied
     end
-
-    return is_shadow_denied
 end
 
-function log_request(lua_metadata, handle)
+function log_request(lua_metadata, handle, rbac_action)
     local client_name = lua_metadata["request.info.client_name"] or ""
     local trusted_client = lua_metadata["request.info.trusted_client"] or false
     local allowed_client = lua_metadata["request.info.allowed_client"] or false
```

Comment:
We can pass it as argument to `log_request`, we already fetched it from metadata